design-proposals/vm-backend-storage.md

```diff
@@ -0,0 +1,73 @@
+Any change a VMI makes outside of its own file-system is lost after a shutdown.  
+Such changes include EFI firmware settings and information stored inside the TPM.  
+This design provides a way to persist small VM-specific files used by virt-launcher.
+
+## Motivation
+Though we now support exposing an emulated TPM device to guests (vTPM), using the device in any meaningful way would require persisting data across reboots.  
+As we do not currently have a way to do that, all data stored by the VMI into the vTPM is lost after a shutdown (a soft reboot keeps the same virt-launcher, avoiding losses).
+
+## Goals
+- Being able to persist modifications of various emulated devices across reboots for a given VM
+- Enable live migration of VMs that use backend storage
+- Avoid consuming large amounts of storage
+- Prevent virt-launcher for VM X from seeing data that belongs to VM Y
+
+## Non Goals
+- Persisting everything. Some information that would usually persist on a non-ephemeral host just don't matter to us.
+- Supporting user modifications of the backend storage. Users whishing to inject their own files should do so at their own risk.
+
+## Definition of Users
+This feature is intended for end-users who wish to preserve the state of some emulated device across reboots.
+
+## User Stories
+As a VM user, I want to be able to store the encryption key of my VMs partitions inside the vTPM (example: BitLocker)
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The storage used for this would be 1 PVC per VM, and would be created on first start of a VM that needs backend storage.  
+This PVC would be RWO, about 2 MiB in size (still TBD), and use [csi-hostpath](https://github.com/kubevirt/hostpath-provisioner).  
+The PVC would be added to the virt-launcher compute container as a volume, but never assigned to the libvirt domain (the guest VM).  
+The PVC would be mounted in a directory at the root of the filesystem. Appropriate symbolic links will be created before starting the libvirt domain.
+The PVC would be owned-by the VM, and therefore will get deleted when its associated VM is deleted.
+
+## API Examples
+Currently, the TPM device API doesn't have any option.  
+We will add a `persistent` boolean option that will default to false.  
+When set to true, the state of the TPM will be persisted.
+
+## Scalability
+Every VM will take a couple more MiBs of storage at most, which shouldn't be a concern, even on cluster running a large number of VMs.  
+For now, the only 2 candidates features for backend storage are:
+- The persistent TPM state, which takes about 6KiB when freshly created. We have yet to find its maximum size.
+- The OVMF vars, which can take up to 528KiB
+I propose finding the biggest size these can take, multiply the sum by 2 and round up to the nearest MiB to allow for future use-cases.
+
+## Update/Rollback Compatibility
+VM migration will have to be specifically handled.
```

Comment:
Does this mean old VMs will get the storage transparently?

Reply:
Yes, if they are modified to use options that leverage backend storage.
Existing VMIs are (mostly) immutable and therefore won't contain any option that require storage.
I can add some language for that.

Reply:
I would appreciate it, thanks!